akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -3726,4 +3726,64 @@ trait FlowOpsMat[+Out, +Mat] extends FlowOps[Out, Mat] {
   def monitor: ReprMat[Out, (Mat, FlowMonitor[Out])] =
     monitorMat(Keep.both)
 
+  /**
+   * This is a more scalable and general case of [[groupedWeightedWithin]]
+   * which groups a stream into vectors based on custom weight and time.
+   * The problem with that solution is each grouped vector must fit into memory before emitting to the next stage.
+   * That won't work for the use case of writing large files.
+   * The desirable behavior is to write data as they come as opposed to accumulate everything until the group slicing condition is met.
+   * In this case, the output channel needs to be closed if there is no data arriving within certain time to avoid connection timeout.
+   * This custom flow uses custom aggregator to support such use cases.
+   * Upstream inputs are continuously aggregated as they arrive.
+   * The currently aggregated elements can be terminated and emitted based on custom conditions
+   *
+   * @param allocate    allocate the initial data structure for aggregated elements
+   * @param aggregate   update the aggregated elements, return true if ready to emit after update
+   * @param emitOnTimer decide whether the currently aggregated elements can be emitted on each timer event
+   * @param harvest     this is invoked before emit within the current stage/operator
+   */
+  def aggregateWithBoundary[Agg, Emit](allocate: => Agg)(
+      aggregate: (Agg, Out) => Boolean,
```

Comment:
There is something important missing here that I didn't notice before (but sort of touched): it must be possible to use this operator also with an immutable data structure but that is currently not possible, since aggregate only returns a boolean, it needs to also return the aggregate data after applying the element.

This means either returning the new agg together with the boolean `(Agg, Boolean)` or making a specific data structure for it (I think I suggested doing an ADT at some time previously in our discussions)

Reply:
The thought is to force `Agg` to be mutable, since using immutable data struct is always inefficient due to allocation. This is kind of radical, but is there potential useful scenario that requires immutable data structure?

Reply:
It is good that we made it possible to use a mutable structure for performance, but it must be possible to use the operator also for immutable data structures (the most obvious use case that comes to mind is for aggregating ByteStrings)

Reply:
fair enough, one more return value in the signature is no big deal.

Reply:
Yeah, will be one tuple allocation, but since it is only called from a single palace, that can hopefully be optimized away by JIT/JVM escape analysis.